File: python/140.py
"""Project Euler Problem 140: Modified Fibonacci golden nuggets.

A_G(x) = x(1+3x)/(1-x-x^2) = N
For rational x, we need 5N^2 + 14N + 1 = k^2 for some integer k
This transforms to the Pell equation X^2 - 5Y^2 = 44 where X = 5N + 7, Y = k
So N = (X - 7)/5, and we need X ≡ 2 (mod 5) for N to be a positive integer
"""

from typing import List, Tuple


def main() -> int:
    """Main function."""
    # All base solutions to X^2 - 5Y^2 = 44
    base_solutions: List[Tuple[int, int]] = [
        (7, 1), (8, 2), (13, 5), (17, 7), (32, 14), (43, 19), (83, 37)
    ]

    all_nuggets: List[int] = []

    for x0, y0 in base_solutions:
        # For the recurrence X_{n+1} = 18*X_n - X_{n-1}, we need X_{-1}
        # We find X_{-1} by working backwards: X_{-1} = 18*X_0 - X_1
        # To find X_1, we use the transformation with fundamental solution (9,4):
        # X_1 = 9*X_0 + 20*Y_0, Y_1 = 4*X_0 + 9*Y_0
        x1 = 9 * x0 + 20 * y0
        y1 = 4 * x0 + 9 * y0
        x_minus1 = 18 * x0 - x1
        y_minus1 = 18 * y0 - y1

        # Check if initial solution gives a nugget
        if x0 > 7 and (x0 - 7) % 5 == 0:
            n = (x0 - 7) // 5
            all_nuggets.append(n)

        # Generate sequence using recurrence
        x_prev, x_curr = x_minus1, x0
        y_prev, y_curr = y_minus1, y0

        # Generate enough terms to ensure we get at least 30 nuggets total
        for _ in range(15):
            x_next = 18 * x_curr - x_prev
            y_next = 18 * y_curr - y_prev

            if x_next > 7 and (x_next - 7) % 5 == 0:
                n = (x_next - 7) // 5
                all_nuggets.append(n)

            x_prev, x_curr = x_curr, x_next
            y_prev, y_curr = y_curr, y_next

    # Remove duplicates, sort and take first 30
    all_nuggets = sorted(set(all_nuggets))
    return sum(all_nuggets[:30])


if __name__ == "__main__":
    print(main())

File: c/140.c
/* Project Euler Problem 140: Modified Fibonacci golden nuggets.
 *
 * Pell equation X^2 - 5Y^2 = 44, need X = 2 (mod 5) for positive integer N.
 * N = (X - 7) / 5.
 * Base solutions and recurrence via fundamental solution (9,4):
 *   X_{n+1} = 9*X_n + 20*Y_n
 *   Y_{n+1} = 4*X_n + 9*Y_n
 *
 * Numbers can get very large, need __int128.
 */
#include <stdio.h>
#include <stdlib.h>

typedef __int128 i128;

/* Print __int128 */
static void print_i128(i128 v) {
    if (v == 0) { printf("0"); return; }
    if (v < 0) { printf("-"); v = -v; }
    char buf[50];
    int pos = 0;
    while (v > 0) {
        buf[pos++] = '0' + (int)(v % 10);
        v /= 10;
    }
    for (int i = pos - 1; i >= 0; i--) putchar(buf[i]);
}

int cmp_i128(const void *a, const void *b) {
    i128 va = *(const i128 *)a;
    i128 vb = *(const i128 *)b;
    if (va < vb) return -1;
    if (va > vb) return 1;
    return 0;
}

int main(void) {
    /* Base solutions to X^2 - 5Y^2 = 44 */
    long long base_x[] = {7, 8, 13, 17, 32, 43, 83};
    long long base_y[] = {1, 2, 5, 7, 14, 19, 37};
    int nbases = 7;

    i128 nuggets[200];
    int nnuggets = 0;

    for (int b = 0; b < nbases; b++) {
        i128 x0 = base_x[b];
        i128 y0 = base_y[b];

        /* Compute x1, y1 using fundamental solution (9, 4) */
        i128 x1 = 9 * x0 + 20 * y0;
        i128 y1 = 4 * x0 + 9 * y0;

        /* Check x0 */
        if (x0 > 7 && (x0 - 7) % 5 == 0) {
            nuggets[nnuggets++] = (x0 - 7) / 5;
        }

        /* Generate using recurrence: X_{n+1} = 18*X_n - X_{n-1} */
        i128 xprev = 18 * x0 - x1; /* x_{-1} */
        i128 xcurr = x0;

        for (int iter = 0; iter < 15; iter++) {
            i128 xnext = 18 * xcurr - xprev;
            if (xnext > 7 && (xnext - 7) % 5 == 0) {
                nuggets[nnuggets++] = (xnext - 7) / 5;
            }
            xprev = xcurr;
            xcurr = xnext;
        }
    }

    /* Sort, deduplicate, take first 30 */
    qsort(nuggets, nnuggets, sizeof(i128), cmp_i128);

    /* Deduplicate */
    int unique = 0;
    for (int i = 0; i < nnuggets; i++) {
        if (i == 0 || nuggets[i] != nuggets[i - 1]) {
            nuggets[unique++] = nuggets[i];
        }
    }

    /* Filter positive only and sum first 30 */
    i128 total = 0;
    int count = 0;
    for (int i = 0; i < unique && count < 30; i++) {
        if (nuggets[i] > 0) {
            total += nuggets[i];
            count++;
        }
    }

    print_i128(total);
    printf("\n");
    return 0;
}
